pkg/kube/secrets/manifest.go

```diff
@@ -0,0 +1,241 @@
+package secrets
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"code.cloudfoundry.org/cf-operator/pkg/bosh/manifest"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+
+	"k8s.io/client-go/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	manifestKeyName   = "manifest"
+	versionKeyName    = "version"
+	sourceKeyName     = "source-description"
+	deploymentKeyName = "deployment-name"
+)
+
+// ManifestPersister is the interface to persist manifests in Kubernetes
+//
+// It uses Kubernetes secrets to persist a manifest. There can only be one
+// manifest per deployment. Each update to the manifest results in a new
+// persisted version. An existing persisted version of a manifest cannot be
+// altered or deleted. The deletion of a manifest will result in the removal
+// of all persisted version of that manifest.
+//
+// The version number is an integer that is incremented with each version of
+// the manifest, which the greatest number being the current/latest version.
+type ManifestPersister interface {
+	// Creates a new secret version of the provided manifest and a description of the "sources" used to render the manifest (e.g. the location of the CRD that generated it)
+	PersistManifest(manifest manifest.Manifest, sourceDescription string) (*corev1.Secret, error)
+
+	// Deletes all persisted versions of the manifest
+	DeleteManifest() error
+
+	// Adds a label to the current/latest version on the manifest
+	DecorateManifest(key string, value string) (*corev1.Secret, error)
+
+	// Lists all persisted versions of the manifest
```

Comment:
I feel like some of these comments are redundant? But it's fine if you want to keep them.